### ISSUE TITLE
Avoid misleading error information for delayed TLS 1.3 alerts

### ISSUE THREAD

**Issue by JeppeTh** [ninenines/gun#341]
When it comes to mTLS TLS 1.3 is not very well designed since the client considers the handshake to be successful before the server has performed it's verification of the client.

If that verification fails it means that the client will receive the TLS alert after ssl:connect has successfully returned.
For such case socket operations like send, setopts etc. will return {error, closed | einval} instead of the tls_alert().
This all leads to that the Gun users get misleading failures. 

This can be solved by letting gun check it's message queue for {ssl_error, Socket, tls_alert()} when failing with closed | einval. 
Needs to be done with a short timeout since the error can be returned before the process receives the alert.

Requires active mode from the start (at least it makes things simpler).

To avoid delay of normal close this can be avoided if TLS 1.3 isn't a candidate, or if the socket has been confirmed (=data received).

**Comment by essen** [comment]
Can you prepare a small example using only the `ssl` application?

**Comment by JeppeTh** [comment]
You meant how to reproduce right? I made some simple example
[delayed_tls_alert.tar.gz](https://github.com/user-attachments/files/17633445/delayed_tls_alert.tar.gz)

Example output:
`48> delayed_tls_alert:run(4004).
delayed_tls_alert:run(4004).
Client successfully connected
send result: ok
send result: ok
send result: ok
Server Handshake result:{error,
                            {tls_alert,
                                {certificate_required,
                                    "TLS server: In state wait_cert at tls_handshake_1_3.erl:1485 generated SERVER ALERT: Fatal - Certificate required\n certificate_required"}}}
send result: {error,closed}
{error,closed}`

**Comment by essen** [comment]
Yes, thank you.

**Comment by essen** [comment]
I don't think we need to do anything about this, as this has been fixed in OTP in https://github.com/erlang/otp/pull/8823

**Comment by JeppeTh** [comment]
That doesn't change the above behaviour. There were situations where the alert could be lost.

ssl:send will never return any alert regardless of that fix.


-------- Original message --------
From: Loïc Hoguin ***@***.***>
Date: 06/11/2024 14:40 (GMT+01:00)
To: ninenines/gun ***@***.***>
Cc: JeppeTh ***@***.***>, Author ***@***.***>
Subject: Re: [ninenines/gun] Avoid misleading error information for delayed TLS 1.3 alerts (Issue #341)


I don't think we need to do anything about this, as this has been fixed in OTP in erlang/otp#8823<https://github.com/erlang/otp/pull/8823>

**Comment by essen** [comment]
Let me see with OTP team if this should be done for more functions.

**Comment by JeppeTh** [comment]
setopts can fail as well at least.


-------- Original message --------
From: Loïc Hoguin ***@***.***>
Date: 06/11/2024 15:46 (GMT+01:00)
To: ninenines/gun ***@***.***>
Cc: JeppeTh ***@***.***>, Author ***@***.***>
Subject: Re: [ninenines/gun] Avoid misleading error information for delayed TLS 1.3 alerts (Issue #341)


Let me see with OTP team if this should be done for more functions.

**Comment by essen** [comment]
We will indeed need to selective receive to check if it's an alert. The alert is now documented in OTP https://github.com/erlang/otp/pull/9067/files

I'm not sure what the timeout should be, perhaps a new option is necessary. I'm also not sure whether we receive the alert message if we close the socket or if we need to wait a bit.

**Comment by essen** [comment]
While I can reproduce with plain `ssl` I am failing to reproduce the same with Gun. With Gun every handshake gets an handshake failure TLS alert and it stops there. It might be related to how Gun starts TLS, it first connect via TCP and then upgrades the connection to TLS.

In any case there's nothing to be done that I can see until we can reproduce with Gun. So I'll be closing this for now. Happy to reopen if there's a way to reproduce.

**Comment by JeppeTh** [comment]
Well I had no problems reproducing with Gun. I would say that it has nothing to do with that gen_tcp is used for connect.
What is required is that mutual TLS 1.3 is used and that it's the server reject the handshake after Gun has replied with its cerificate.

**Comment by essen** [comment]
OK I understand what I did wrong. Let me try again.

**Comment by JeppeTh** [comment]
A bit funny - I tried to change my example to make use of tcp connect followed by ssl handshake - but I can't get it working at all. The handshakes hangs forever - something obvious I guess but I fail to spot it :)

But now when looking at Gun it does ssl:connect immediately?

**Comment by essen** [comment]
OK double checking everything, I do get the following eventually, after the handshake went through successfully:

```
    [{gun_down,<0.165.0>,http,
         {error,
             {tls_alert,
                 {certificate_required,
                     "TLS client: In state connection received UNDEFINED ALERT: Fatal - Certificate required\n"}}},
         []}]}
```

But this comes up from active mode. What I think you describe is sending when active mode is not currently active inside Gun, which will be a little trickier to reproduce in a normal test. But when modifying Gun to not set active mode, then Gun doesn't receive the alert at all, and all I have is the `{error, closed}`.

> But now when looking at Gun it does ssl:connect immediately?

This was changed in 2.0 (I think), now Gun first connects via TCP and then does the TLS handshake, to have separate events.

**Comment by essen** [comment]
This is the rough patch against current master with a WIP test and small changes to Gun to try to trigger this:

```
diff --git a/src/gun.erl b/src/gun.erl
index 085e47f..506b0d9 100644
--- a/src/gun.erl
+++ b/src/gun.erl
@@ -1696,7 +1696,7 @@ active(State=#state{active=false}) ->
 active(State=#state{transport=gun_quicer}) ->
 	{ok, State};
 active(State=#state{socket=Socket, transport=Transport}) ->
-	case Transport:setopts(Socket, [{active, once}]) of
+	case Transport:setopts(Socket, [{active, false}]) of
 		ok ->
 			{ok, State};
 		{error, closed} ->
diff --git a/src/gun_http.erl b/src/gun_http.erl
index d28600b..930993a 100644
--- a/src/gun_http.erl
+++ b/src/gun_http.erl
@@ -654,6 +654,8 @@ send_request(State=#http_state{socket=Socket, transport=Transport, version=Versi
 	SendResult = Transport:send(Socket, [
 		cow_http:request(Method, Path, Version, Headers),
 		[Body || Body =/= undefined]]),
+
+							timer:sleep(1000), ct:pal("g ~p", [process_info(self(), messages)]),
 	EvHandlerState2 = EvHandler:request_headers(RequestEvent, EvHandlerState1),
 	EvHandlerState = case Out of
 		head ->
@@ -730,12 +732,15 @@ data(State=#http_state{socket=Socket, transport=Transport, version=Version,
 								EvHandlerState0),
 							{{state, State#http_state{out=head}}, EvHandlerState};
 						Error={error, _} ->
+							timer:sleep(100), ct:pal("g ~p", [process_info(self(), messages)]),
 							{Error, EvHandlerState0}
 					end;
 				body_chunked when Version =:= 'HTTP/1.1' ->
 					case Transport:send(Socket, cow_http_te:chunk(Data)) of
 						ok -> {[], EvHandlerState0};
-						Error={error, _} -> {Error, EvHandlerState0}
+						Error={error, _} -> 
+							timer:sleep(100), ct:pal("g ~p", [process_info(self(), messages)]),
+						{Error, EvHandlerState0}
 					end;
 				{body, Length} when DataLength =< Length ->
 					Length2 = Length - DataLength,
@@ -750,6 +755,7 @@ data(State=#http_state{socket=Socket, transport=Transport, version=Version,
 						ok when Length2 > 0, IsFin =:= nofin ->
 							{{state, State#http_state{out={body, Length2}}}, EvHandlerState0};
 						Error={error, _} ->
+							timer:sleep(100), ct:pal("g ~p", [process_info(self(), messages)]),
 							{Error, EvHandlerState0}
 					end;
 				body_chunked -> %% HTTP/1.0
diff --git a/test/gun_SUITE.erl b/test/gun_SUITE.erl
index 1c9ba23..e661b27 100644
--- a/test/gun_SUITE.erl
+++ b/test/gun_SUITE.erl
@@ -633,6 +633,39 @@ supervise_false(_) ->
 	[] = [P || {_, P, _, _} <- supervisor:which_children(gun_sup), P =:= Pid],
 	ok.
 
+tls_alert_delayed(_) ->
+	%% @todo doc
+	{ok, ListenSocket} = ssl:listen(0, [binary, {active, true},
+		{versions, ['tlsv1.3']},
+		%% The alert will be triggered by the missing certificate.
+		{verify, verify_peer},
+		{fail_if_no_peer_cert, true},
+		{keyfile, "/home/essen/Downloads/server.key"},
+		{certfile, "/home/essen/Downloads/server.crt"},
+		{cacertfile, "/home/essen/Downloads/cacert.pem"}]),
+	{ok, {_, OriginPort}} = ssl:sockname(ListenSocket),
+	AcceptPid = spawn_link(fun() ->
+		{ok, ClientSocket} = ssl:transport_accept(ListenSocket, 5000),
+		{error, {tls_alert, _}} = ssl:handshake(ClientSocket, 5000),
+		receive after infinity -> ok end
+	end),
+	{ok, ConnPid} = gun:open("localhost", OriginPort, #{
+		retry => 0,
+		transport => tls,
+		tls_opts => [{verify, verify_none}]
+		%,trace => true
+	}),
+	ct:pal("~p", [gun:await_up(ConnPid)]),
+	timer:sleep(1000),
+	StreamRef = gun:post(ConnPid, "/", [{<<"content-type">>, <<"application/octet-stream">>}]),
+%	spawn_link(fun SendLoop() -> gun:data(ConnPid, StreamRef, nofin, rand:bytes(64)), timer:sleep(10), SendLoop() end),
+	timer:sleep(2000),
+	ct:pal("~p", [process_info(self(), messages)]),
+	gun:close(ConnPid).
+
+
+
+
 tls_handshake_error_gun_http2_init_retry_0(_) ->
 	doc("Ensure an early TLS connection close is propagated "
 		"to the user of the connection."),

**Comment by JeppeTh** [comment]
You managed to reproduce now?

**Comment by essen** [comment]
I get the unhelpful `{error, closed}` but I don't get the TLS alert as a message.

**Comment by JeppeTh** [comment]
I believe you must enable active mode from start - i.e. as a connect option - otherwise possible a risk that it is missed.

**Comment by essen** [comment]
Right but the way things work in Gun now is:

* passive TCP socket gets opened
* then socket gets upgraded to TLS
* then active mode gets enabled

And when I try I always get an error returned when enabling active mode (returned by `setopts`), and that error is always the alert.

Gun 2.x doesn't do, like in the example, connect then send, it's connect (in two steps) then setopts then send, and setopts does return the alert. So it's hard to get a test case where the alert would not be caught by setopts.

I wonder if we can set `fail_if_no_peer_cert` to `false` initially, then request/require the certificate some time after the connection is established. In that scenario it would be easier to get it to have the alert on send.

**Comment by JeppeTh** [comment]
Ok - now I get it - gen_tcp:connect followed by ssl:connect is how you upgrade the client.
For my private testing with this I added [{active,true}] in ssl:connect and I never experience losing the alert.
I now realize there was a bug in OTP that the alert could be lost - perhaps you run without that.

fail_if_no_peer_cert is a server option - and I don't think it can be used for delaying the handshake if it was mainly for testing.

**Comment by JeppeTh** [comment]
[delayed_tls_alert_upgrade.zip](https://github.com/user-attachments/files/19492731/delayed_tls_alert_upgrade.zip)
At least here's an upgrade example where at least I can reproduce. Can see if I can try it out with Gun in some way later

**Comment by essen** [comment]
Your example works because you didn't disable active mode when connecting. The default is active mode:

```
3> {ok, S} = gen_tcp:connect("google.com", 80, []).
{ok,#Port<0.8>}
4> inet:getopts(S, [active]).
{ok,[{active,true}]}
```

So your `gen_tcp:connect` needs to set `{active, false}` to mirror what Gun does.

Once you do that, it will likely be more difficult to reproduce, although I didn't try it on your example yet.

**Comment by JeppeTh** [comment]
But there's really no reason for gun to avoid active mode?

Anyhow, I have no problem reproducing if adding {active,false} in the connect calls.

The thing about lost alert was fixed in OTP-26.2.5.4 - but sounds like you rather always get the alert from ssl:setopts.
I'm currently running OTP-26.2.5.5.

**Comment by essen** [comment]
It does work with your example, I just now tested it.

It however does not work if you change your example to add a `timer:sleep(1)` just before calling `setopts`. If you add an `io:format` to print the `setopts` error, you get the same result I can see with Gun.

Gun does more work than this example, and while it doesn't explicitly sleep, it probably gets scheduled out, which has a similar effect. And so the likelihood of triggering the alert before `setopts` has completed is smaller. I have not been able to reproduce with Gun so far.

Since it's dependent on timing, it also depends on environment. Clearly my environment doesn't help with that.

So I am wondering what I could do to make it more likely. One thing I am thinking now is to use the `cb_info` option to introduce a delay every time the server-side `ssl` sends. This should help ensure that the alert gets sent after Gun has called setopts. Then we can make Gun send data to mirror what's happening in the example.

I'll try it tomorrow, it's late. Thanks for bearing with me. 👍

**Comment by JeppeTh** [comment]
Here's a variant with a proxy in between otherwise - then it should be simple to simulate a slow network ;)
[delayed_with_proxy.zip](https://github.com/user-attachments/files/19495931/delayed_with_proxy.zip)

**Comment by essen** [comment]
I still can't have a test using the current Gun version that reproduces. Delaying data doesn't seem to help, but perhaps I'm not cutting the data right. But the handshake in Gun doesn't return until the last piece of data is sent, so I'm not sure what could be different.

You can find the current test at https://github.com/ninenines/gun/pull/new/tls_alert_delayed and run it with `make ct-gun c=tls_alert_delayed`. Please try to modify it until you succeed in reproducing.

I tried on OTP-26.2.5.9 and OTP-27.2.4. I don't know what I can do to make it miss the TLS alert.

**Comment by JeppeTh** [comment]
Not sure if I got blind chasing the wrong type of fault etc...
[gun_SUITE.zip](https://github.com/user-attachments/files/19512278/gun_SUITE.zip)

It seems alpn h2 makes things differently. With above update I can reproduce the fault.
I.e. the TLS alert is missed and instead you get {error,{down,{shutdown,closed}}}

There was some issue I believe in case failure is detected ahead of connected or not. When I'm running the above test it seems its Protocol:init that fails. Then I need to check for a delayed alert and if so swap the "closed" reason.

But in my prior testing in our system also ssl:negotiated_protocol could lead to the same thing.  (closed or einval in all cases)
So also here this "maybe delayed tls alert" was needed.

Then both the above could succeed and then instead while sending or activating etc operation could fail. And then instead disconnect with close is called for instance. So also in disconnect this "maybe delayed tls alert" operation is needed. I.e. check if there's a tls alert in message queue.

Then I'm not sure if needed, but in our system testing we always enabled active mode in ssl:connect to ensure no alert was missed when toggling active mode.

Here's another variant of the suite but with the proxy where things perhaps can be delayed simpler. But not sure if it's needed.

[gun_SUITE_proxy.zip](https://github.com/user-attachments/files/19512530/gun_SUITE_proxy.zip)

**Comment by essen** [comment]
Thanks! Enabling ALPN on the server does change things and I also get the unhelpful error in that case.

**Comment by essen** [comment]
In that case though the Gun process does not receive the TLS alert. :)

**Comment by essen** [comment]
I think it comes down to this. If we have active mode at or before this point: https://github.com/ninenines/gun/blob/master/src/gun.erl#L1338 - then we get the alert error.

I think it can be enabled in the `gun_tls:connect` call, given via options. I don't know if that's a good idea yet, though.

**Comment by essen** [comment]
It seems that trying to enable active mode after the ssl:send has failed returns the error.

```
{error,
    {tls_alert,
        {certificate_required,
            "TLS client: In state connection received UNDEFINED ALERT: Fatal - Certificate required\n"}}}
```

So this is a better way than blindly receiving the message IMO. Just need to do that after negotiated_protocol call and after the ssl:send in HTTP/2, I think.

**Comment by JeppeTh** [comment]
Even if you can't manage to reproduce it in your test, einval can for sure happen. It's based on timing I believe.

Why not keep it simple and always check if there's an alert regardless of the error reason?

I also believe there's a risk that the alert has to be waited for, at least in some of my testing it could be lost unless waiting for 100ms. 
Hence why I also added some socket confirmed flag in case application data has been received on the socket.

I can't understand how they could approve the TLS 1.3 spec - a nightmare to manage this "maybe verified connection"

**Comment by essen** [comment]
I think `einval` cannot happen in Gun 2.x. In Gun 1.x not all socket operations would be checked for errors, so for example if a send is done and then setopts, and only the setopts is checked for error, then setopts could return `einval` as long as send returned `closed`. But because we check all errors in Gun 2.x then we will get the `closed` when it happens and stop there. `einval` is returned only after the `closed` error, so we shouldn't get it anymore (and if we do we can fix wherever an error was ignored).

If the connection was closed as a result of the alert, then when we call setopts after receiving a `closed` ssl will immediately give us the alert. So I don't think there's any waiting necessary there.

Where we might miss it is after active mode has been enabled, if the alert gets sent to the Gun process and Gun gets a `closed` error and disconnects. When disconnecting Gun will flush all pending socket messages, including the alert. So I think that's the only place where a receive might be necessary. But I don't know if that can happen in Gun 2.0 yet as even if I do delays I get the error in the HTTP/2 send or when setting active mode in HTTP/1. But I will try a few more things to see if I can make it happen in other places.

So at this point I prefer handling it where it's been observed to happen with the current Gun version. If there's still a case that needs fixing it can be further improved later on.

**Comment by essen** [comment]
Having post-handshake auth would help to write a test case where the alert gets sent when active mode has been enabled, so I have opened a ticket: https://github.com/erlang/otp/issues/9667

**Comment by essen** [comment]
I have tests for HTTP/1 and HTTP/2, with the alert coming from different locations. I note that depending on when I run the tests in the test suite, I get the alert where I expect it, or later, and in the case of HTTP/1 I can get the alert after active mode has been enabled. So if I can isolate exactly what creates this delay I may be able to get a good enough test case that has a send error while active mode is enabled.

I noticed a discrepancy in the message being forwarded (one is wrapped in an additional error tuple, the other not) but that's a "problem" to be solved at a later time and not directly related to this so I left a todo.

I double checked whether we can get any TLS alert data from an `ssl:negotiated_protocol` returning `{error, closed}` and we can't as the error is only returned when the process is already gone. Since we are not in active mode at that point, there's not much we can do. But that's anomalous as we just completed the handshake successfully, so the TLS process shouldn't just be gone without giving us an error at that point. If the process is still alive we will get the error when we call the next ssl function.

I have also left a todo for HTTP/3 but while QUIC uses TLS 1.3 it has its own implementation so the details for QUIC will need to be figured out independently. Not now though.

**Comment by essen** [comment]
Bingo!

```
gun_SUITE:tls13_post_handshake_alert_http1 failed on line 673
Reason: {badmatch,{error,{down,{shutdown,{error,closed}}}}}
```

**Comment by JeppeTh** [comment]
`double checked whether we can get any TLS alert data from an ssl:negotiated_protocol returning {error, closed} and we can't as the error is only returned when the process is already gone.`

What do you mean by this? It's called in the same process as connect? If so it's bug in OTP or you're running an OTP version without a fix for that.

**Comment by essen** [comment]
No, I mean that by reading the code of `ssl:negotiated_protocol`, we can see that when the ssl process exists, it always returns a good value (the protocol negotiated or the error that no protocol was negotiated), and that `{error, closed}` is only returned when the ssl process exits (caller process, Gun, got an exit signal, via the `gen:call` mechanism). If the process is gone, for whatever reason, the alert is also gone, since we don't have active mode at that point. But I don't see this happening except under really heavy load or through cascading failures, at which point you have other things to worry about.

**Comment by essen** [comment]
You can't just enable active mode, ssl:negotiated_protocol, like the HTTP/2 send, must be handled specially. So I will need a test case if there's an issue with it. Not that I'm saying it's impossible, but I need to see how it works and confirm we do forward the alert properly. I'll do a few more tries to trigger it, I have an idea or two.

The `einval` happens when sending because we get the raw error from the underlying TCP socket's send. Unlike with TCP sockets it doesn't necessarily mean that the TLS "socket" (process) is gone which is interesting to say the least. After more testing I can indeed confirm that we can get the alert from the socket even though send returns `einval`, via the setopts method.

```
h1 send data error {error,einval}
followed by setopts {error,
                                   {tls_alert,
                                    {certificate_required,
                                     "TLS client: In state connection received UNDEFINED ALERT: Fatal - Certificate required\n"}}}
```

So we can handle alerts on both `closed` and `einval` with a detailed comment about the latter as it's not obvious.

**Comment by JeppeTh** [comment]
We have used gun_tls:connect with active mode for several months without noticing any drawbacks for our use cases.

**Comment by essen** [comment]
I'm talking about the upcoming Gun 2.2 with the changes resulting from this ticket. I will push them to master most likely tomorrow, I'm doing more test runs to see if I can make it fail in ssl:negotiated_protocol but that hasn't happened once in many runs so Gun 2.2 will likely not do anything special for ssl:negotiated_protocol.

**Comment by JeppeTh** [comment]
Ok, that is unfortunate if so since it can for sure happen.

**Comment by essen** [comment]
What I am seeing is the following, with a timer:sleep (1ms to 1s, random) before calling `ssl:negotiated_protocol`:

* active mode disabled: `ssl:negotiated_protocol` never fails. Instead a subsequent call to setopts/send fails and the alert can be retrieved then
* active mode enabled: `ssl:negotiated_protocol` easily fails with the timer:sleep and the alert is in the messages Gun received

But since at that point Gun has active mode disabled, we are in the first scenario, so it looks like it cannot fail, but also that Gun will not lose the alert since it'll be retrieved via the next calls.

I think that you can see it happen because you enabled active mode, but that won't be necessary (and will be harmful) in Gun 2.2.

**Comment by essen** [comment]
It works exactly the same when using HTTP/1.1 CONNECT to a TLS server from a clear-text TCP proxy. We are in passive mode when that happens because when we receive the CONNECT response the `{active,once}` mode gets disabled automatically, and active mode is enabled only after the TLS upgrade. So there's nothing special to be done there.

HTTP/1.1 CONNECT in a TLS over TLS, as well as HTTP/2 CONNECT, still remain to be tested.

**Comment by JeppeTh** [comment]
Ok - we'll see if our testing will find any glitches later on.

**Comment by essen** [comment]
Testing alerts with tunnels is reliable on Linux but on macOS/Windows often the alert is simply not available, at all, regardless of active mode. I don't know why, I would expect the alert sent at least when we are in active mode. I will disable the alert tests for tunnels on macOS/Windows for now.

**Comment by JeppeTh** [comment]
Great - don't you just love the TLS 1.3 design 😉

Let me see when I can find the time to try out this in our testing. I believe we're running Gun 2.1, not sure if some of the changes requires 2.2...

**Comment by JeppeTh** [comment]
It was fairly simple for me to still reproduce missed alerts. I believe your maybe_tls_alert should match on tls_alert errors and not just any error.

I have situations where the message in the queue is
`{ssl_error,
         {sslsocket,
             {gen_tcp,#Port<0.13736>,tls_connection,undefined},
             [<0.132594.0>,<0.132593.0>]},
         closed}`

Then a bit later this pops up that we really are waiting for
`{ssl_error,
         {sslsocket,
             {gen_tcp,#Port<0.13736>,tls_connection,undefined},
             [<0.132594.0>,<0.132593.0>]},
         {tls_alert,
             {certificate_required,
                 "TLS client: In state connection received SERVER ALERT: Fatal - Certificate required\n"}}}`

**Comment by JeppeTh** [comment]
I don't have my old test cases remaining, had to make some new which possibly are a bit different. 
Anyhow another problem I see is that the tls alert is detected between await_up and set_owner which means my test fails with noproc instead of the alert.  But I guess that is not related to your patch - I will try later with my own patch to compare.

**Comment by JeppeTh** [comment]
I made a fix/workaround for the OTP bug and then Gun seems rock solid with the above fix.

**Comment by essen** [comment]
Sounds great! I'll include the fix in the next Gun release.

Please submit the details to OTP if it's something they can fix, or detail it here otherwise, I'm curious. :-)

**Comment by JeppeTh** [comment]
I'm in contact with OTP but not sure they will fix it.  
The problem is on the server side kind of. The client/OTP at least fails to receive the TLS alert from the server.
And the reason seam to be like

1. Server sends TLS alert + FIN
2. Client send magic
3. Server send FIN+ACK
4. Client sends http2 headers  of the first request
5. Server responds with RST

In the above scenario it seems the TLS Alert doesn't reach the OTP client.
OTP server has some kind of protection to avoid the TLS Alert to be lost. They simple send the alert and than wait for something to be received from the other side before shutdown.

My workaround/suggested solution is that instead of just wait for anything wait for the client to close the connection. At least that solved my test cases so that I could verify your changes more thoroughly.

**Comment by JeppeTh** [comment]
Seems OTP doesn't like that gun_http2:init sends stuff in passive mode which they believe is a reason for the missed the alert.
They were a bit busy so I couldn't get all details.  I'm questioning it since active mode should not be required - but perhaps it's something about mixing passive and active once.

Anyhow I'm making some testing now just to see if it makes a difference or not.

I first tried active once during tls:connect, and then suddenly ssl:negotiated_protocol started to fail - so that probably explains why I saw that with my patch but not with yours (I had active enabled in ssl:connect)

Moved active once ahead of proto init instead and will see what happens

**Comment by essen** [comment]
Yes ssl:negotiated_protocol only fails if active mode is enabled. But if it's disabled then it is the subsequent setopts that fail where ssl:negotiated_protocol would have.

**Comment by essen** [comment]
As long as the ssl process is alive I don't see a good reason for us to miss the alert just because we are in passive mode.

**Comment by JeppeTh** [comment]
I'm guessing they mean that passive mode will miss that socket is closed and thereby next send will happen and lead to RST.
I.e. I'm guess they are trying to avoid the extra send...

**Comment by JeppeTh** [comment]
Eventually I managed to get the "lost alert due to RST" even if active one before magic is sent.

**Comment by JeppeTh** [comment]
> It was fairly simple for me to still reproduce missed alerts. I believe your maybe_tls_alert should match on tls_alert errors and not just any error.

Any plan to include this correction?

**Comment by essen** [comment]
For future self we're talking about https://github.com/ninenines/gun/issues/341#issuecomment-2807006794

I'll see when I can include it. I have many fixes coming up for Gun, probably can't get everything ready for September's release though.

**Comment by essen** [comment]
I did not have time to add half of what I wanted to Gun unfortunately. There's always next month. Thanks for your patience.
